// payload/dist/fields/hooks/afterRead/relationshipPopulationPromise.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const types_1 = require("../../config/types");
const populate = async ({ depth, currentDepth, req, overrideAccess, dataReference, data, field, index, showHiddenFields, }) => {
    const dataToUpdate = dataReference;
    const relation = Array.isArray(field.relationTo) ? data.relationTo : field.relationTo;
    const relatedCollection = req.payload.collections[relation];
    if (relatedCollection) {
        let idString = Array.isArray(field.relationTo) ? data.value : data;
        if (typeof idString !== 'string' && typeof (idString === null || idString === void 0 ? void 0 : idString.toString) === 'function') {
            idString = idString.toString();
        }
        let populatedRelationship;
        if (depth && currentDepth <= depth) {
            populatedRelationship = await req.payload.findByID({
                req,
                collection: relatedCollection.config.slug,
                id: idString,
                currentDepth: currentDepth + 1,
                overrideAccess: typeof overrideAccess === 'undefined' ? false : overrideAccess,
                disableErrors: true,
                depth,
                showHiddenFields,
            });
        }
        // If populatedRelationship comes back, update value
        if (populatedRelationship || populatedRelationship === null) {
            if (typeof index === 'number') {
                if (Array.isArray(field.relationTo)) {
                    dataToUpdate[field.name][index].value = populatedRelationship;
                }
                else {
                    dataToUpdate[field.name][index] = populatedRelationship;
                }
            }
            else if (Array.isArray(field.relationTo)) {
                dataToUpdate[field.name].value = populatedRelationship;
            }
            else {
                dataToUpdate[field.name] = populatedRelationship;
            }
        }
    }
};
const relationshipPopulationPromise = async ({ siblingDoc, field, depth, currentDepth, req, overrideAccess, showHiddenFields, }) => {
    const resultingDoc = siblingDoc;
    const populateDepth = (0, types_1.fieldHasMaxDepth)(field) && field.maxDepth < depth ? field.maxDepth : depth;
    if ((0, types_1.fieldSupportsMany)(field) && field.hasMany && Array.isArray(siblingDoc[field.name])) {
        const rowPromises = [];
        siblingDoc[field.name].forEach((relatedDoc, index) => {
            const rowPromise = async () => {
                if (relatedDoc) {
                    await populate({
                        depth: populateDepth,
                        currentDepth,
                        req,
                        overrideAccess,
                        data: relatedDoc,
                        dataReference: resultingDoc,
                        field,
                        index,
                        showHiddenFields,
                    });
                }
            };
            rowPromises.push(rowPromise());
        });
        await Promise.all(rowPromises);
    }
    else if (siblingDoc[field.name]) {
        await populate({
            depth: populateDepth,
            currentDepth,
            req,
            overrideAccess,
            dataReference: resultingDoc,
            data: siblingDoc[field.name],
            field,
            showHiddenFields,
        });
    }
};
exports.default = relationshipPopulationPromise;
